### PR TITLE
CI and Tests stabilization

### PR DESCRIPTION
**Pull request by lidel**
This PR aims to smoke-tests potential fixes for:

- Jenkins: init's "before all" hook sometimes timeouts #370
- Random DNS problems during Docker build #371 

(ignore it for now, as I plan to rebase it until Jenkins and tests are 👌)

### CODE REVIEW

**Comment by lidel** [comment]
Ok, it seems to work quite good so far:

# Overview

- isolated build within docker
  - glibc-based image 
     (solves #371 + enables us to run stuff from http://dist.ipfs.io in test suite in future)
  - built-in cleanup (removal of images and containers after the build)
  - supports parallel builds (docker image and container names are scoped per branch)
- `ci:*` scripts can be run manually (local testing, reproducible builds during review etc)
- removed arrow functions and increased mocha timeout to 15s  (did not experience #370  since)
- added addon linter at the end (it lints build output, not sources)
- test results are visualized on "Tests" tab
- `.zip` with extension can be downloaded from "Artifacts" tab

# Screenshot 

screen shot 2018-02-07 at 23 18 15

File: ci/Jenkinsfile [hunk 1]
```diff
@@ -1,7 +1,30 @@
...
+        checkout scm
+        sh 'printenv'
+    }
+    stage('ci:install') {
+        sh 'docker build -t ipfs-companion:${JOB_BASE_NAME}-${BUILD_NUMBER} .'
+    }
+    catchError {
+        stage('ci:test') {
+            sh 'docker run -i --name ipfs-companion-test-${JOB_BASE_NAME}-${BUILD_NUMBER} -e JUNIT_REPORT_PATH=test/report.xml ipfs-companion:${JOB_BASE_NAME}-${BUILD_NUMBER} npm run ci:test'
+            sh 'docker cp ipfs-companion-test-${JOB_BASE_NAME}-${BUILD_NUMBER}:/usr/src/app/test/report.xml test/report.xml || true' // copy test report
```
Comment (lidel) [inline ci/Jenkinsfile]: `|| true'` is a workaround for false-negative bug in docker (more details @ https://github.com/moby/moby/issues/3986)

**Comment by VictorBjelkholm** [comment]
It is not obvious in the "blue ocean" UI (the one the status check points to) but there is a merge conflict that stops the jenkins build from starting: https://ci.ipfs.team/job/IPFS%20Shipyard/job/ipfs-companion/job/PR-372/31/console

**Comment by lidel** [comment]
Ok, resolved conflicts and merged changes from master.  
I see there is still something fishy about `docker cp` when I copy directory tree (chown error that I work around by `|| true`). 
It seems to be the last thing blocking this PR. I'll look into it tomorrow.

**Comment by lidel** [comment]
@VictorBjelkholm  `cleanWs()` solved permission issues + removed risk of subsequent builds impacting each other. 

Any last concerns before I merge this?

**Pull request merged**